lib/pavilion/builder.py

```diff
@@ -1014,3 +1011,84 @@ def __eq__(self, other):
         compares = [getattr(self, key) == getattr(other, key)
                     for key in compare_keys]
         return all(compares)
+
+def get_used_build_paths(tests_dir: Path) -> set:
+
+    used_builds = set()
+
+    for path in dir_db.select(tests_dir):
+        build_origin_symlink = path/'build_origin'
+        build_origin = None
+        if (build_origin_symlink.exists() and
+            build_origin_symlink.is_symlink() and
+            utils.resolve_path(build_origin_symlink).exists()):
+            build_origin = build_origin_symlink.resolve()
+
+        if build_origin is not None:
+            used_builds.add(path.name)
+
+    return used_builds
+
+def delete_unused(tests_dir: Path, builds_dir: Path, verbose: bool = False) -> int:
+    """Delete all the build directories, that satisfy some filter.
+
+    :param tests_dir: The test_runs directory path object.
+    :param builds_dir: The builds directory path object.
+    :param args: The parsed command arguments object.
+
+    :return int count: The number of builds that were removed.
+
+    """
+
+    used_build_paths = get_used_build_paths(tests_dir)
+
+    def filter_builds(path: Path) -> bool:
+        """Build filter function. Build paths that have no associated 
+        tests and can be removed return true, otherwise they return false.
+
+        :param path: The passes build path object
+
+        :return True: The passed build path can be removed.
+        :return False: The passed build path cannot be removed.
+
+        """
+        return not path.name in used_build_paths
+
+    count = 0
+
+    lock_path = builds_dir.with_suffix('.lock;')
+    error_list = []
+    with lockfile.LockFile(lock_path):
+        for path in dir_db.select(builds_dir, filter_builds, fn_base = 16):
+            try:
+                shutil.rmtree(path.as_posix())
+                path.with_suffix('.finished').unlink()
+            except OSError as err:
+                error_list.append("Could not remove build {}: {}"
+                                  .format(path, err))
+                continue
+            count += 1
+            if verbose:
+                output.fprint('Removed build {}.'.format(path.name))
+
+    return count, error_list
+
+def remove_lingering_finished_files(builds_dir: Path) -> None:
```

Comment:
I think we can do without this. You remove them above when you remove the build. I just need to change how new builds handle a missing build when .finished exists.